src/components/v5/common/Fields/Datepicker/Datepicker.tsx

```diff
@@ -0,0 +1,180 @@
+import clsx from 'clsx';
+import isAfter from 'date-fns/isAfter';
+import React, { useCallback, useRef, useState } from 'react';
+import DatePicker from 'react-datepicker';
+import 'react-datepicker/dist/react-datepicker.css';
+
+import { useMobile } from '~hooks';
+import { formatText } from '~utils/intl.ts';
+import Button from '~v5/shared/Button/index.ts';
+import { type ButtonMode } from '~v5/shared/Button/types.ts';
+
+import { DATEPICKER_PORTAL_ID, DEFAULT_DATE_FORMAT } from './consts.ts';
+import DatepickerContainer from './partials/DatepickerContainer/index.ts';
+import DatepickerCustomHeader from './partials/DatepickerCustomHeader/index.ts';
+import DatepickerCustomInput from './partials/DatepickerCustomInput/index.ts';
+import DatepickerCustomRangeInput from './partials/DatepickerCustomRangeInput/index.ts';
+import { type DatepickerProps } from './types.ts';
+
+import styles from './Datepicker.css';
+
+const Datepicker = <WithRange extends boolean = false>({
+  cancelButtonProps,
+  applyButtonProps,
+  popperModifiers,
+  selectsRange,
+  onChange,
+  startDate: startDateProp,
+  endDate: endDateProp,
+  selected: selectedProp,
+  dateFormat = DEFAULT_DATE_FORMAT,
+  popperClassName,
+  ...rest
+}: DatepickerProps<WithRange>): ReturnType<React.FC> => {
+  const isMobile = useMobile();
+  const calendarRef = useRef<DatePicker<never, WithRange>>(null);
+
+  const [startDate, setStartDate] = useState<Date | null>(
+    startDateProp || null,
+  );
+  const [endDate, setEndDate] = useState<Date | null>(endDateProp || null);
+
+  const handleChange: DatepickerProps<WithRange>['onChange'] = useCallback(
+    (dates) => {
+      if (!dates && selectsRange) {
+        setStartDate(null);
+        setEndDate(null);
+        return;
+      }
+
+      if (Array.isArray(dates)) {
+        const [start, end] = dates;
+        setStartDate(start);
+        setEndDate(end);
+      } else {
+        setStartDate(dates);
+      }
+    },
+    [selectsRange],
+  );
+
+  const resetValues = useCallback(() => {
+    if (selectsRange) {
+      setStartDate(startDateProp || null);
+      setEndDate(endDateProp || null);
+    } else {
+      setStartDate(selectedProp || null);
+      setEndDate(null);
+    }
+  }, [endDateProp, selectedProp, selectsRange, startDateProp]);
```

Comment:
The props passed to these callbacks are _by designed_ made to be changed every single time the specific function is called, so memorizing them makes no sense in this context.

They should just be simple functions, as they don't require all that overhead.